### PR TITLE
Bug Fixes and Race Controls

### PR DESCRIPTION
**Pull request by Jakllp**
- Fixed some bugs (see commit notes)
- Race Controls HAD some problems
- Can now drive when standing on edge of block
- Got rid of "empty commits"

### CODE REVIEW

File: src/main/resources/plugin.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 name: uCars
 main: com.useful.ucars.ucars
-version: 21
+version: 21.5
```
Comment (storm345) [inline src/main/resources/plugin.yml:3]: 22 please :P

File: src/main/java/com/useful/ucars/uCarsListener.java [hunk 2]
```diff
@@ -562,13 +568,15 @@ public void tickCalcsAndLegacy(VehicleUpdateEvent event) {
 				UEntityMeta.removeMetadata(vehicle, "car.vec"); //Update the 'car.vec' metadata with an otherwise identical event; but without the player object attached
 				ucarUpdateEvent et = new ucarUpdateEvent(vehicle, evt.getTravelVector().clone(), null, evt.getDir()); //Clone of the other event, except no player object attached
 				et.setRead(evt.getReadCount()); //Make sure it IS a clone (With correct variable values)
+				if(vehicle.hasMetadata("car.frozen")) {
```
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:571]: Why is this required? The uCarUpdateEvent I think should still be fired for a frozen car

File: src/main/java/com/useful/ucars/uCarsListener.java [hunk 3]
```diff
@@ -1275,20 +1293,13 @@ public void run() {
...
-																		// step
-																		// block
-																		// and
-																		// trying
-																		// to
-																		// jump
+				if (carBlock.name().toLowerCase().contains("slab")) { // In a slab block
 					calculated = true;
-					y = 0.6;
+					y = 1.2;
```
Comment (storm345) [inline src/main/java/com/useful/ucars/uCarsListener.java:1302]: Does this definitely not work as 0.6? Would this being 1.2 not make the car float too high when driving over a road made of half slabs?

**Review by storm345 — CHANGES_REQUESTED** [review]
See comments on the code

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 3 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: I've just tried again and the lowest I can go is 1.1 (without going into 2 decimal digits)
Lower and you can not get of a slab-road.
You float as high as a normal block, yes.
It is actually similar to 0.6 float but here is the catch:
As soon as I get up to a block I drop.
This could have to do with some other jumping mechanics, I'll check.

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: Hmm you should be able to be offset 0.6 when inside the slab, and then if you are trying to drive off of the road (into a full size block) detect that and then decide to climb - not be permanently too high for the slab

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: Yes I'm probably gonna implement something like that.
Maybe the 0.6 offset wont even be necessary after that as the floating looks a bit odd...

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: The 0.6 is to calculate what height it should be at I think, eg. block y coordinate+0.6 (which should be just above the top of the slab)

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 2 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:571]: This is currently needed as cars that should be frozen can drive when you leave this out.

I'm gonna look into placing this check somewhere else

Comment (storm345) [reply src/main/java/com/useful/ucars/uCarsListener.java:571]: okay, in the update event would make more sense I think. There should have already been a check somewhere?

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:571]: There is one in calculateCarStats which is called during the update event.
The thing is that (at least with race controls) this does not work.
I'm gonna look into it

File: src/main/java/com/useful/ucars/uCarsListener.java (hunk 3 above)
Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: Okay so I added that check in a slightly different way. Climbing and stuff works now BUT there is a Problem:

The slab elevator will make you float basically half a block over the slab.
What is that slab elevator even supposed to achieve?

Comment (Jakllp) [reply src/main/java/com/useful/ucars/uCarsListener.java:1302]: Nevermind that one, figured it out, will fix

**Comment by Jakllp** [comment]
Fixed (at least I think so)

**Comment by storm345** [comment]
Looks like it's probably fine now. Have you tested with driving on slabs, double slabs, etc... and it all worked fine?

**Comment by Jakllp** [comment]
Seems to work perfectly fine

**Comment by storm345** [comment]
Okay cool will merge it!
